### PR TITLE
Only init variable names to nil for output element variable

### LINKED ISSUE #4100: Improve nil initialisation of locally scoped outputElement for feel expressions
For multi instance activities we can define the outputElement as a feel expression. To make sure that this variable is available during execution of the multi instance body, it needs to be initialised with a nil-value. Otherwise, each iteration over the body might write values to it at a non-local scope. This only needs to be done for expressions that are just a variable or a nested variable.

This was quickly (and ugly) coded with a simple regex check. But that might lead to some unnecessary (i.e. unwanted) variables on the local scope. For example, when a more advanced expression is used as an outputElement.

This is specifically for `engine/processor/workflow/handlers/multiinstance/AbstractMultiInstanceBodyHandler.java`

### PR DESCRIPTION
**Pull request by korthout**
## Description

Introduced a `Optional<String> getVariableName()` method to `Expression` that provides a name of the top most variable in the expression if it is a single variable or a nested property of a variable.

This method is in turn used to init the output element variable to nil at the local scope of a multi instance body, to make sure such a body can set a variable with that name at its own scope.

## Related issues

closes #4100

### CODE REVIEW

File: expression-language/src/main/java/io/zeebe/el/Expression.java [hunk 1]
```diff
@@ -7,12 +7,20 @@
...
 public interface Expression {
 
   /** @return the (raw) expression as string */
   String getExpression();
 
+  /**
+   * @return optional of the name of the variable if expression is a single variable or a property
+   *     of a single variable, otherwise empty
+   */
+  Optional<String> getVariableName();
```
Comment (saig0) [inline expression-language/src/main/java/io/zeebe/el/Expression.java:22]: Please add some unit tests for the different implementations.

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java [hunk 2]
```diff
@@ -518,6 +521,48 @@ public void shouldIterateOverNestedInputCollection() {
...
+        ENGINE
+            .workflowInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariable(INPUT_COLLECTION_EXPRESSION, INPUT_COLLECTION)
+            .create();
+
+    RecordingExporter.jobRecords(JobIntent.CREATED)
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .limit(INPUT_COLLECTION.size())
+        .exists();
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java:545]: This check is done in `completeJobs()`

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java [hunk 3]
```diff
@@ -518,6 +521,48 @@ public void shouldIterateOverNestedInputCollection() {
...
+    final List<Long> scopeKeys =
+        RecordingExporter.variableRecords(VariableIntent.CREATED)
+            .withName(OUTPUT_ELEMENT_EXPRESSION)
+            .map(Record::getValue)
+            .map(VariableRecordValue::getScopeKey)
+            .limit(3)
+            .collect(Collectors.toList());
+    assertThat(scopeKeys)
+        .hasSize(3)
+        .allSatisfy(scopeKey -> assertThat(scopeKeys).containsOnlyOnce(scopeKey));
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java:563]: What do you want to check here?

I think that we should verify that 
* the variable is created without `.nested` in its name (which is the issue that was fixed)
* the variable is initialized with `null`
* the values are collected in the output collection

**Review by saig0 — CHANGES_REQUESTED** [review]
I like it :+1: 

Just a few comments about the tests.

**Comment by korthout** [comment]
@saig0 Please have another look

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java [hunk 4]
```diff
@@ -539,28 +539,67 @@ public void shouldCollectNestedOutputElements() {
...
+  public void shouldCollectOutputElementsFromExpression() {
+    // given
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            workflow(
+                miBuilder.andThen(
+                    m ->
+                        m.zeebeOutputElementExpression(
+                            "number(string(loopCounter) + string(loopCounter))"))))
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java:575]: That looks tricky :joy:

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java [hunk 5]
```diff
@@ -539,28 +539,67 @@ public void shouldCollectNestedOutputElements() {
...
+
+    // complete jobs
+    completeJobs(workflowInstanceKey, INPUT_COLLECTION.size());
+
+    // then
+    assertThat(
+            RecordingExporter.variableRecords(VariableIntent.CREATED)
+                .map(Record::getValue)
+                .map(VariableRecordValue::getName)
+                .limit(10))
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java:593]: Limiting the records to a specific number is a bit unsteady. Instead, we can limit the records from creating a workflow instance to its completion/termination.

```suggestion
             RecordingExporter.records()
                .limitToWorkflowInstance(workflowInstanceKey)
                .variableRecords()
                .map(Record::getValue)
                .map(VariableRecordValue::getName))
```

**Review by saig0 — APPROVED** [review]
:+1: Just one improvement for the test.

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java (hunk 5 above)
Comment (korthout) [reply engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java:593]: Nice, that's a good one to learn about.

**Pull request merged**